adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala

```diff
@@ -311,25 +340,30 @@ sealed abstract class AlignmentRecordRDD extends AvroReadGroupGenomicRDD[Alignme
    * @return CoverageRDD containing mapped RDD of Coverage.
    */
   def toCoverage(): CoverageRDD = {
-    val covCounts =
-      rdd.rdd
-        .filter(r => {
-          val readMapped = r.getReadMapped
-
-          // validate alignment fields
-          if (readMapped) {
-            require(r.getStart != null && r.getEnd != null && r.getContigName != null,
-              "Read was mapped but was missing alignment start/end/contig (%s).".format(r))
-          }
-
-          readMapped
-        }).flatMap(r => {
-          val t: List[Long] = List.range(r.getStart, r.getEnd)
-          t.map(n => (ReferenceRegion(r.getContigName, n, n + 1), 1))
-        }).reduceByKey(_ + _)
-        .map(r => Coverage(r._1, r._2.toDouble))
-
-    RDDBoundCoverageRDD(covCounts, sequences, None)
+    import dataset.sqlContext.implicits._
+    val covCounts = dataset.toDF
+      .where($"readMapped")
+      .select($"contigName", $"start", $"end")
+      .as[AlignmentWindow]
+      .flatMap(w => {
+        val width = (w.end - w.start).toInt
+        val buffer = new Array[Coverage](width)
+        var idx = 0
```

Comment:
It would be more concise to do the following:
`val positions = Array.range(w.start, w.end)`
`positions.map(f => Coverage(w.contigname, f, f + 1L, 1.0))`